python/semantic_kernel/memory/protocols/memory_record_base.py

```diff
@@ -0,0 +1,43 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+
+from abc import ABC, abstractmethod
+from typing import Generic, TypeVar
+
+DataModelT = TypeVar("DataModelT", bound=object)
+KeyTypeT = TypeVar("KeyTypeT")
+
+
+class MemoryRecordServiceBase(ABC, Generic[DataModelT, KeyTypeT]):
+    def __init__(self, item_type: DataModelT, key_type: KeyTypeT | None = None):
+        if not hasattr(item_type, "__sk_data_model__"):
+            raise ValueError(f"Item type {item_type} must be a data model, use the @datamodel decorator")
+        self._item_type: DataModelT = item_type
+        if not key_type:
+            self._key_type: KeyTypeT = item_type.__sk_data_model_fields__.key_field
+        else:
+            self._key_type: KeyTypeT = key_type
+
+    @abstractmethod
+    async def upsert(self, collection_name: str, record: DataModelT, **kwargs) -> KeyTypeT:
```

Comment:
FYI, for collection_name, we have made it so that it can be passed either into the constructor or into the method, but it's optional on both, and you have to pass it into at least one.  The one passed to the constructor is essentially an optional default, and the one in the method is an optional override.